include/modem/at_parser.h

```diff
@@ -0,0 +1,307 @@
+	AT_PARSER_CMD_TYPE_READ,
+	/** AT test command. */
+	AT_PARSER_CMD_TYPE_TEST
+};
+
+/**
+ * @brief AT parser
+ *
+ * Holds the parsing state for a configured AT command string.
+ *
+ */
+struct at_parser {
+	/* Pointer to the current AT command line. */
+	const char *at;
+	/* Pointer to the current AT command line value. */
+	const char *cursor;
+	/* Number of values parsed so far for the current AT command line. */
+	size_t count;
+	/* The next subparameter is empty. */
+	bool is_next_empty;
+	/* Sentinel value set at initialization. */
+	uint32_t init_sentinel;
+};
+
+/**
+ * @brief Type-generic macro for getting a subparameter value as an integer.
+ *
+ * @param[in]  parser AT parser.
+ * @param[in]  index  Subparameter index in the current AT command line configured in @p parser.
+ * @param[out] value  Subparameter value.
+ *
+ * @retval 0 If the operation was successful.
+ *           Otherwise, a (negative) error code is returned.
+ * @retval -EINVAL     One or more of the supplied parameters are invalid.
+ * @retval -EPERM      @ref parser has not been initialized.
+ * @retval -EOPNOTSUPP Operation not supported for the subparameter type at the given index.
+ * @retval -ERANGE     Parsed integer value is out of range for the expected type.
+ * @retval -EBADMSG    The AT command string is malformed.
+ */
+#define at_parser_num_get(parser, index, value)    \
+	_Generic((value),                          \
+		int16_t * : at_parser_int16_get,   \
+		uint16_t * : at_parser_uint16_get, \
+		int32_t * : at_parser_int32_get,   \
+		uint32_t * : at_parser_uint32_get, \
+		int64_t * : at_parser_int64_get,   \
+		uint64_t * : at_parser_uint64_get  \
+	)(parser, index, value)
+
+/**
+ * @brief Initialize an AT parser for a given AT command string.
+ *
+ * @param[in] parser A pointer to the AT parser.
+ * @param[in] at     A pointer to the AT command string to parse.
+ *
+ * @retval 0 If the operation was successful.
+ *           Otherwise, a (negative) error code is returned.
+ * @retval -EINVAL One or more of the supplied parameters are invalid.
+ */
+int at_parser_init(struct at_parser *parser, const char *at);
+
+/**
+ * @brief Move the cursor of an AT parser to the next line of its configured AT command string.
+ *
+ * It is possible to call this function successfully with a given AT parser if and only if its
+ * configured AT command string has multiple lines and its current line is not the last
+ * one.
+ *
+ * @param[in] parser A pointer to the AT parser.
+ *
+ * @retval 0 If the operation was successful.
+ *           Otherwise, a (negative) error code is returned.
+ * @retval -EINVAL     One or more of the supplied parameters are invalid.
+ * @retval -EPERM      @p parser has not been initialized.
+ * @retval -EOPNOTSUPP The cursor cannot move to the next line. This error occurs if there is no
+ *                     next line to move to or the configured AT command string is malformed.
+ */
+int at_parser_next(struct at_parser *parser);
```

Comment:
Perhaps `_line_next()` would be a bit more descriptive. It's not clear that this function moves the parser to the next line without reading the documentation.

Reply:
Updated.